[night-issues: Cycle Verify] Maven: sitemanage

## Cycle Verify residual (next-cycle lead)

**not a qa task.** Unassigned implementation residual for the overnight lead queue.

### Integration tip
- `integration_ref`: `423138e4c2aa6a14a2271a1a314094a18623e08c` (`feat/issue-3796-http-json-source-chrome`, PR #3800)
- Cluster was skipped; newest this-run WebUI/QA PR used as the integration tip.

### Command
```
cd projects/sitemanage
..\..\mvnw.cmd clean install
```
Worktree: `%USERPROFILE%\.grok\worktrees\intersoft-workspace-percussioncms\night-issue-prs`

### First failure lines
```
[ERROR] Failures:
[ERROR]   ContentTypeAdaptorWorkflowsTest.put_conflictWhenLockNotHeld:219 Expected com.percussion.rest.contenttypes.ContentTypeDesignLockException to be thrown, but nothing was thrown.
[ERROR]   ContentTypeAdaptorWorkflowsTest.put_conflictWhenLockedByOtherUser:237 Expected com.percussion.rest.contenttypes.ContentTypeDesignLockException to be thrown, but nothing was thrown.
[ERROR]   ContentTypeAdaptorWorkflowsTest.put_conflictWhenLockedInAnotherSession:257 Expected com.percussion.rest.contenttypes.ContentTypeDesignLockException to be thrown, but nothing was thrown.
[ERROR]   ContentTypeAdaptorWorkflowsTest.put_emptyListClearsAssociations:166 Wanted but not invoked: pSContentEditor.setWorkflowInfo(...)
[ERROR]   ContentTypeAdaptorWorkflowsTest.put_rejectsUnknownWorkflowId:206 Expected java.lang.IllegalArgumentException to be thrown, but nothing was thrown.
[ERROR]   ContentTypeAdaptorWorkflowsTest.put_rejectsUnknownWorkflowName:180 Expected java.lang.IllegalArgumentException to be thrown, but nothing was thrown.
[ERROR] Errors:
[ERROR]   ContentTypeAdaptorWorkflowsTest.get_listsNewSetAfterPut:149 NullPointer Cannot invoke "ContentTypeDetail.getAllowedWorkflows()" because "put" is null
[ERROR]   ContentTypeAdaptorWorkflowsTest.put_persistsWorkflowsWhenLockHeld:132 NullPointer Cannot invoke "ContentTypeDetail.getAllowedWorkflows()" because "out" is null
[ERROR] Tests run: 1503, Failures: 6, Errors: 2, Skipped: 125
[ERROR] Failed to execute goal maven-surefire-plugin:3.5.4:test (default-test) on project sitemanage: There are test failures.
```

Work #3795 noted the same 8 `ContentTypeAdaptorWorkflowsTest` failures as present on `origin/main` without that PR's diff. Cycle Verify reproduced them on the integration tip (main + PR #3800 WebUI chrome).

### This-run PR URLs
- https://github.com/intersoftdatalabs-in/percussioncms/pull/3798 (merged; system http-json SPI)
- https://github.com/intersoftdatalabs-in/percussioncms/pull/3799 (rest persist http-json)
- https://github.com/intersoftdatalabs-in/percussioncms/pull/3800 (WebUI http-json chrome — integration tip)
- https://github.com/intersoftdatalabs-in/percussioncms/pull/3801 (content type item exits REST)
- https://github.com/intersoftdatalabs-in/percussioncms/pull/3804 (content type control property values REST)

### Other gates this cycle
- Maven green on integration tip: `system` (Tests run: 2413, Failures: 0, Skipped: 244), `rest` (572/0), `WebUI` (Java 63/0; Vitest 392 files / 3072 passed), `modules/perc-qa-automation` (BUILD SUCCESS).
- Playwright green: golden-unattended-smoke (2), login (2), developer-site-virtual-source (20).

### Acceptance
- `cd projects/sitemanage && ../../mvnw.cmd clean install` (or `../mvnw` on Unix) is **BUILD SUCCESS** with **Failures: 0, Errors: 0** for `ContentTypeAdaptorWorkflowsTest` (and the rest of the sitemanage surefire suite).
- not a qa task

> Co-Authored by Grok Build 1.0.5 using grok-4.6 with agent night-issue-prs-cycle-verify.

## Agent progress (night-issue-prs)

| Slice | Issue | Status | PR | Notes | Updated |
|-------|-------|--------|----|-------|---------|
| CD-08 workflows Maven residual | #3805 | pr_opened | [#3812](https://github.com/intersoftdatalabs-in/percussioncms/pull/3812) | `setAllowedWorkflows` item-def existence + `reloadItemDef` cache-miss fallback; sitemanage clean install Tests run: 1535, Failures: 0, Errors: 0 | 2026-08-25 |

--- a/docs/ai-generated/code-reviews/3805-content-type-adaptor-workflows-erlang.md
+++ b/docs/ai-generated/code-reviews/3805-content-type-adaptor-workflows-erlang.md
@@ -1,0 +1,46 @@
+# Erlang review — issue #3805 ContentTypeAdaptor CD-08 workflows
+
+**Branch:** `fix/issue-3805-content-type-adaptor-workflows`  
+**Date:** 2026-08-25  
+**Recommendation:** approve  
+**Gate:** May commit/push: yes  
+**Memory patterns hit:** change-class closure (bugfix of existing CD-08 adaptor; rest companions already on main); behavioral test for post-save cache miss (not structural-only); no path I/O; `resolveItemDef` after PUT save must not 404 a committed persist (CD-07 peer).
+
+## Summary
+
+`ContentTypeAdaptor.setAllowedWorkflows` (CD-08) now matches `setContentTypeEnabled` for existence (item-def lookup so unknown ids 404 before lock) and matches CD-07 `replaceFieldControlProperties` for post-save reload (`reloadItemDef` + fall back to the locked definition). A post-save `PSInvalidContentTypeException` is no longer swallowed as HTTP 404 after a successful `saveContentTypes`.
+
+Cycle Verify (#3805) reported 6 failures + 2 errors in `ContentTypeAdaptorWorkflowsTest` on the full sitemanage suite: lock 409 not thrown, unknown workflow 400 not thrown, empty-list `setWorkflowInfo` never invoked, PUT returning null. Those symptoms match an early null return (existence via `loadContentTypes(lock=false)`) and/or `resolveItemDef` throwing after save.
+
+## Issues
+
+None blocking.
+
+Existence via `resolveItemDef` is the same pattern as `setContentTypeEnabled` (suite-green peer). Post-save `reloadItemDef` is the CD-07 fix already approved on this class. `put_cacheMissAfterSave_fallsBackToLockedDef` stubs `getItemDef(311L)` as first-return / second-throw so existence succeeds and reload misses.
+
+Lock / unknown-workflow / empty-list / persist tests remain and now also assert allowed-workflow list size on GET after PUT.
+
+No public/protected signature change. No rest module change. Product-docs already describe `200` + `allowedWorkflows` (CD-08); this restores that contract.
+
+## Cross-platform path checklist
+
+Applied. No filesystem path construction. REST path strings in existing product-docs use URI `/`.
+
+## Tests
+
+- `put_cacheMissAfterSave_fallsBackToLockedDef` — behavioral coverage of cache-miss fallback
+- Persist / GET round-trip list size restored
+- Existing 409 lock / 400 unknown id and name / empty-list clear / 404 unknown type / 403 / session still present
+- Standalone `cd projects/sitemanage && ../../mvnw.cmd clean install`: **BUILD SUCCESS**, Tests run: 1535, Failures: 0, Errors: 0, Skipped: 125; `ContentTypeAdaptorWorkflowsTest` 13/0/0
+
+## Non-blocking notes (not issues)
+
+- `setContentTypeEnabled` still calls `resolveItemDef` after save (same pre-existing cache-miss 404). Out of scope for #3805.
+- Empty `loadContentTypes(lock=true)` after a held lock still returns null (404). Same as enable/disable; do not treat as a CD-08 gate.
+
+## Handoff
+
+- Reviewed: uncommitted issue #3805 work in worktree `night-issue-prs`
+- Top finding: none blocking
+- Recommendation: **approve**. May commit/push: **yes**.
+- Artifact: `docs/ai-generated/code-reviews/3805-content-type-adaptor-workflows-erlang.md`

--- a/projects/sitemanage/src/main/java/com/percussion/apibridge/ContentTypeAdaptor.java
+++ b/projects/sitemanage/src/main/java/com/percussion/apibridge/ContentTypeAdaptor.java
@@ -637,14 +637,16 @@ public class ContentTypeAdaptor implements IContentTypesAdaptor {
     if (trimmed.contains("*")) {
       throw new IllegalArgumentException("idOrName must not contain wildcards");
     }
+    requireSessionUserForLock();
     String session = currentSession();
     String user = currentUser();
-    requireSessionUserForLock();
     try {
-      IPSGuid ctGuid = resolveExistingContentTypeGuid(trimmed);
-      if (ctGuid == null) {
+      // Peer setContentTypeEnabled: item-def existence so unknown ids 404 before lock.
+      PSItemDefinition current = resolveItemDef(trimmed);
+      if (current == null) {
         return null;
       }
+      IPSGuid ctGuid = new PSGuid(PSTypeEnum.NODEDEF, current.getTypeId());
       requireHeldLock(ctGuid);
       List<PSItemDefinition> locked;
       try {
@@ -676,7 +678,8 @@ public class ContentTypeAdaptor implements IContentTypesAdaptor {
             "Failed to save content type workflow associations {}: {}", idOrName, e.getMessage(), e);
         throw new IllegalStateException("Failed to save content type workflow associations", e);
       }
-      PSItemDefinition reloaded = resolveItemDef(trimmed);
+      // Cache miss after save must not 404 a successful persist (CD-07 peer).
+      PSItemDefinition reloaded = reloadItemDef(trimmed);
       return reloaded != null ? toDetail(reloaded) : toDetail(def);
     } catch (ContentTypeDesignLockException
         | IllegalArgumentException

--- a/projects/sitemanage/src/test/java/com/percussion/apibridge/ContentTypeAdaptorWorkflowsTest.java
+++ b/projects/sitemanage/src/test/java/com/percussion/apibridge/ContentTypeAdaptorWorkflowsTest.java
@@ -131,6 +131,7 @@ class ContentTypeAdaptorWorkflowsTest {
         adaptor.setAllowedWorkflows(null, "311", List.of(simple, standard), simple);
 
     assertNotNull(out);
+    assertEquals(2, out.getAllowedWorkflows().size());
     assertEquals("Simple Workflow", out.getDefaultWorkflow().getName());
     verify(designWs).saveContentTypes(anyList(), eq(false), eq("test-session"), eq("Admin"));
     ArgumentCaptor<PSWorkflowInfo> info = ArgumentCaptor.forClass(PSWorkflowInfo.class);
@@ -148,12 +149,16 @@ class ContentTypeAdaptorWorkflowsTest {
     NamedObjectRef simple = namedWorkflow("Simple Workflow", 4);
     ContentTypeDetail put = adaptor.setAllowedWorkflows(null, "311", List.of(simple), simple);
     assertNotNull(put);
+    assertEquals(1, put.getAllowedWorkflows().size());
+    assertEquals("Simple Workflow", put.getAllowedWorkflows().get(0).getName());
     ArgumentCaptor<PSWorkflowInfo> info = ArgumentCaptor.forClass(PSWorkflowInfo.class);
     verify(editor).setWorkflowInfo(info.capture());
     assertEquals(List.of(4), workflowIds(info.getValue()));
 
     ContentTypeDetail get = adaptor.getContentType(null, "311");
     assertEquals("percPage", get.getName());
+    assertEquals(1, get.getAllowedWorkflows().size());
+    assertEquals("Simple Workflow", get.getAllowedWorkflows().get(0).getName());
     assertEquals("Simple Workflow", get.getDefaultWorkflow().getName());
   }
 
@@ -164,10 +169,29 @@ class ContentTypeAdaptorWorkflowsTest {
 
     ContentTypeDetail out = adaptor.setAllowedWorkflows(null, "311", List.of(), null);
 
+    assertNotNull(out);
     ArgumentCaptor<PSWorkflowInfo> info = ArgumentCaptor.forClass(PSWorkflowInfo.class);
     verify(editor).setWorkflowInfo(info.capture());
     assertTrue(workflowIds(info.getValue()).isEmpty());
     assertTrue(out.getAllowedWorkflows().isEmpty());
+  }
+
+  @Test
+  void put_cacheMissAfterSave_fallsBackToLockedDef() throws Exception {
+    stubHeldLock();
+    PSItemDefinition def = stubDefinition();
+    // Existence uses item-def; post-save reload may miss the cache (CD-07 peer).
+    when(itemDefManager.getItemDef(eq(311L), eq(PSItemDefManager.COMMUNITY_ANY)))
+        .thenReturn(def)
+        .thenThrow(new PSInvalidContentTypeException("not cached"));
+
+    NamedObjectRef simple = namedWorkflow("Simple Workflow", 4);
+    ContentTypeDetail out = adaptor.setAllowedWorkflows(null, "311", List.of(simple), simple);
+
+    verify(designWs).saveContentTypes(anyList(), eq(false), eq("test-session"), eq("Admin"));
+    assertNotNull(out);
+    assertEquals("Simple Workflow", out.getDefaultWorkflow().getName());
+    verify(def.getContentEditor()).setWorkflowId(4);
   }
 
   @Test

